

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>fiber_views package &mdash; fiber-views 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=2709fde1"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Examples" href="examples.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            fiber-views
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Install:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Install</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">fiber_views package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#fiber-views-annotation-module">fiber_views.annotation module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#fiber-views-cli-module">fiber_views.cli module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#fiber-views-fiber-views-module">fiber_views.fiber_views module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.fiber_views.FiberView"><code class="docutils literal notranslate"><span class="pre">FiberView</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.fiber_views.FiberView.get_seq_records"><code class="docutils literal notranslate"><span class="pre">FiberView.get_seq_records()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.fiber_views.FiberView.mark_cpg_sites"><code class="docutils literal notranslate"><span class="pre">FiberView.mark_cpg_sites()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.fiber_views.build_fview"><code class="docutils literal notranslate"><span class="pre">build_fview()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.fiber_views.build_multi_fview"><code class="docutils literal notranslate"><span class="pre">build_multi_fview()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.fiber_views.build_single_fview"><code class="docutils literal notranslate"><span class="pre">build_single_fview()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fiber-views-plot-module">fiber_views.plot module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.annotate_boundaries"><code class="docutils literal notranslate"><span class="pre">annotate_boundaries()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.draw_fiber_bars"><code class="docutils literal notranslate"><span class="pre">draw_fiber_bars()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.draw_fiber_lines"><code class="docutils literal notranslate"><span class="pre">draw_fiber_lines()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.draw_mods"><code class="docutils literal notranslate"><span class="pre">draw_mods()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.draw_mods_offset"><code class="docutils literal notranslate"><span class="pre">draw_mods_offset()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.draw_regions"><code class="docutils literal notranslate"><span class="pre">draw_regions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.draw_split_lines"><code class="docutils literal notranslate"><span class="pre">draw_split_lines()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.plot.make_plot_ax"><code class="docutils literal notranslate"><span class="pre">make_plot_ax()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fiber-views-tools-module">fiber_views.tools module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.agg_by_obs_and_bin"><code class="docutils literal notranslate"><span class="pre">agg_by_obs_and_bin()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.bin_sparse_regions"><code class="docutils literal notranslate"><span class="pre">bin_sparse_regions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.calc_kmer_dist"><code class="docutils literal notranslate"><span class="pre">calc_kmer_dist()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.count_kmers"><code class="docutils literal notranslate"><span class="pre">count_kmers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.filter_regions"><code class="docutils literal notranslate"><span class="pre">filter_regions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.get_seq_records"><code class="docutils literal notranslate"><span class="pre">get_seq_records()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.get_sequences"><code class="docutils literal notranslate"><span class="pre">get_sequences()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.make_dense_regions"><code class="docutils literal notranslate"><span class="pre">make_dense_regions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.make_region_df"><code class="docutils literal notranslate"><span class="pre">make_region_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.mark_cpg_sites"><code class="docutils literal notranslate"><span class="pre">mark_cpg_sites()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.plot_methylation"><code class="docutils literal notranslate"><span class="pre">plot_methylation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.plot_summary"><code class="docutils literal notranslate"><span class="pre">plot_summary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.simple_region_plot"><code class="docutils literal notranslate"><span class="pre">simple_region_plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.tools.split_fire"><code class="docutils literal notranslate"><span class="pre">split_fire()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fiber-views-utils-module">fiber_views.utils module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.ReadList"><code class="docutils literal notranslate"><span class="pre">ReadList</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.strand"><code class="docutils literal notranslate"><span class="pre">ReadList.strand</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.build_anno_df"><code class="docutils literal notranslate"><span class="pre">ReadList.build_anno_df()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.build_mod_array"><code class="docutils literal notranslate"><span class="pre">ReadList.build_mod_array()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.build_mod_array_from_def"><code class="docutils literal notranslate"><span class="pre">ReadList.build_mod_array_from_def()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.build_seq_array"><code class="docutils literal notranslate"><span class="pre">ReadList.build_seq_array()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.build_sparse_region_array"><code class="docutils literal notranslate"><span class="pre">ReadList.build_sparse_region_array()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.filter_by_end_meth"><code class="docutils literal notranslate"><span class="pre">ReadList.filter_by_end_meth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.filter_by_window"><code class="docutils literal notranslate"><span class="pre">ReadList.filter_by_window()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.get_reads"><code class="docutils literal notranslate"><span class="pre">ReadList.get_reads()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#fiber_views.utils.ReadList.print_aligned_centers"><code class="docutils literal notranslate"><span class="pre">ReadList.print_aligned_centers()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.bed_to_anno_df"><code class="docutils literal notranslate"><span class="pre">bed_to_anno_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.get_mod_pos_from_rec"><code class="docutils literal notranslate"><span class="pre">get_mod_pos_from_rec()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.get_strand_correct_mods"><code class="docutils literal notranslate"><span class="pre">get_strand_correct_mods()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.get_strand_correct_mods_from_def"><code class="docutils literal notranslate"><span class="pre">get_strand_correct_mods_from_def()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.get_strand_correct_regions"><code class="docutils literal notranslate"><span class="pre">get_strand_correct_regions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.make_sparse_regions"><code class="docutils literal notranslate"><span class="pre">make_sparse_regions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.print_mod_contexts"><code class="docutils literal notranslate"><span class="pre">print_mod_contexts()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fiber_views.utils.read_bed"><code class="docutils literal notranslate"><span class="pre">read_bed()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-contents">Module contents</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">fiber-views</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">fiber_views package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/fiber_views.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="fiber-views-package">
<h1>fiber_views package<a class="headerlink" href="#fiber-views-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="fiber-views-annotation-module">
<h2>fiber_views.annotation module<a class="headerlink" href="#fiber-views-annotation-module" title="Link to this heading"></a></h2>
<p id="module-fiber_views.annotation">Created on Mon Mar  6 12:03:38 2023</p>
<p>&#64;author: morgan</p>
</section>
<section id="fiber-views-cli-module">
<h2>fiber_views.cli module<a class="headerlink" href="#fiber-views-cli-module" title="Link to this heading"></a></h2>
<p id="module-fiber_views.cli">Console script for fiber_views.</p>
</section>
<section id="fiber-views-fiber-views-module">
<h2>fiber_views.fiber_views module<a class="headerlink" href="#fiber-views-fiber-views-module" title="Link to this heading"></a></h2>
<p id="module-fiber_views.fiber_views">Main module.</p>
<dl class="py class">
<dt class="sig sig-object py" id="fiber_views.fiber_views.FiberView">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fiber_views.fiber_views.</span></span><span class="sig-name descname"><span class="pre">FiberView</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-1000,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_mod_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">220</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mark_cpgs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fully_span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'cutoff':</span> <span class="pre">2,</span> <span class="pre">'dist':</span> <span class="pre">3000}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['np',</span> <span class="pre">'ec',</span> <span class="pre">'rq']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fire</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.fiber_views.FiberView" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.fiber_views.FiberView.get_seq_records">
<span class="sig-name descname"><span class="pre">get_seq_records</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'read_name'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.fiber_views.FiberView.get_seq_records" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.fiber_views.FiberView.mark_cpg_sites">
<span class="sig-name descname"><span class="pre">mark_cpg_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sparse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.fiber_views.FiberView.mark_cpg_sites" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.fiber_views.build_fview">
<span class="sig-prename descclassname"><span class="pre">fiber_views.fiber_views.</span></span><span class="sig-name descname"><span class="pre">build_fview</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-1000,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fully_span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'cutoff':</span> <span class="pre">2,</span> <span class="pre">'dist':</span> <span class="pre">3000}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['np',</span> <span class="pre">'ec',</span> <span class="pre">'rq']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fire</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.fiber_views.build_fview" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.fiber_views.build_multi_fview">
<span class="sig-prename descclassname"><span class="pre">fiber_views.fiber_views.</span></span><span class="sig-name descname"><span class="pre">build_multi_fview</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_defs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_defs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-1000,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fully_span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'cutoff':</span> <span class="pre">2,</span> <span class="pre">'dist':</span> <span class="pre">3000}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['np',</span> <span class="pre">'ec',</span> <span class="pre">'rq']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.fiber_views.build_multi_fview" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.fiber_views.build_single_fview">
<span class="sig-prename descclassname"><span class="pre">fiber_views.fiber_views.</span></span><span class="sig-name descname"><span class="pre">build_single_fview</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_info</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_defs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_defs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-1000,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fully_span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'cutoff':</span> <span class="pre">2,</span> <span class="pre">'dist':</span> <span class="pre">3000}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['np',</span> <span class="pre">'ec',</span> <span class="pre">'rq']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_reads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.fiber_views.build_single_fview" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="fiber-views-plot-module">
<h2>fiber_views.plot module<a class="headerlink" href="#fiber-views-plot-module" title="Link to this heading"></a></h2>
<p id="module-fiber_views.plot">Created on Wed Aug 28 13:33:42 2024</p>
<p>&#64;author: morgan</p>
<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.annotate_boundaries">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">annotate_boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.annotate_boundaries" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.draw_fiber_bars">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">draw_fiber_bars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#d0d0d0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.draw_fiber_bars" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.draw_fiber_lines">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">draw_fiber_lines</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#606060'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.draw_fiber_lines" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.draw_mods">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">draw_mods</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'m6a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#000000'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.draw_mods" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.draw_mods_offset">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">draw_mods_offset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'m6a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#000000'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.draw_mods_offset" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.draw_regions">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">draw_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'msp'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.draw_regions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.draw_split_lines">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">draw_split_lines</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site_name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.draw_split_lines" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.plot.make_plot_ax">
<span class="sig-prename descclassname"><span class="pre">fiber_views.plot.</span></span><span class="sig-name descname"><span class="pre">make_plot_ax</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.plot.make_plot_ax" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="fiber-views-tools-module">
<h2>fiber_views.tools module<a class="headerlink" href="#fiber-views-tools-module" title="Link to this heading"></a></h2>
<p id="module-fiber_views.tools">Created on Wed Sep  7 14:37:04 2022</p>
<p>&#64;author: morgan</p>
<p>&#64;description: A set of usefull tools for workign with fiber views</p>
<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.agg_by_obs_and_bin">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">agg_by_obs_and_bin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_group_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site_name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_to_keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['seqid',</span> <span class="pre">'pos',</span> <span class="pre">'strand',</span> <span class="pre">'']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ones'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.agg_by_obs_and_bin" title="Link to this definition"></a></dt>
<dd><p>Aggregate fiber view data by a group variable in the <cite>obs</cite> dataframe and bin by <cite>bin_widht</cite> basepairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – The fiber view object containing the data to be aggregated.</p></li>
<li><p><strong>obs_group_var</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the <cite>obs</cite> group variable to use for aggregation. The default value is ‘site_name’.
If <cite>obs_group_var</cite> is set to None, the fiber view will not be aggregated by rows and the row ordering will
be preserved.</p></li>
<li><p><strong>bin_width</strong> (<em>int</em><em>, </em><em>optional</em>) – The width of each bin, in base pairs. The default value is 10.
If <a href="#id1"><span class="problematic" id="id2">`</span></a>bin_width is 1, the data will not be binned.</p></li>
<li><p><strong>obs_to_keep</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – A list of observation metadata columns to keep in the aggregated data. The default value is [‘seqid’, ‘pos’, ‘strand’, ‘’].</p></li>
<li><p><strong>fast</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the modification matrices will be converted to dense matrices for faster calculations. The default value is True.
This may use more memory for large fiber view objects.</p></li>
<li><p><strong>region_weights</strong> (<em>str</em><em>, </em><em>optional</em>) – how to weight regions when aggregating must be one of ‘ones’, ‘length’ or ‘score’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An aggregated version of the input fiber view object, with observations grouped and binned according to the specified parameters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.bin_sparse_regions">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">bin_sparse_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nuc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.bin_sparse_regions" title="Link to this definition"></a></dt>
<dd><p>Bin regions in a fiber view by averaging their length and score over a set of consecutive bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – The fiber view object containing the region data.</p></li>
<li><p><strong>base_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the type of regions to bin. This should be one of ‘nuc’ (nucleosomes), or ‘msp’ (methylation sensitive patches).
The default value is ‘nuc’.</p></li>
<li><p><strong>bin_width</strong> (<em>int</em><em>, </em><em>optional</em>) – The width of each bin, in base pairs. The default value is 10.</p></li>
<li><p><strong>interval</strong> (<em>int</em><em>, </em><em>optional</em>) – The interval between bins, in base pairs. The default value is 3.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the position, length, and score data for the binned regions, stored as
COOrdinate format sparse matrices.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of scipy.sparse.coo_matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.calc_kmer_dist">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">calc_kmer_dist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cityblock'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.calc_kmer_dist" title="Link to this definition"></a></dt>
<dd><p>Calculate pairwise k-mer distances between fibers in a fiber view.</p>
<p>This function calculates pairwise distances between fibers in a fiber view based on the k-mer counts stored in the ‘kmers’ element of the <cite>obsm</cite> attribute. The distance metric can be specified using the <cite>metric</cite> parameter (default is ‘cityblock’). The resulting distance matrix is stored in the ‘kmer_dist’ element of the <cite>obsp</cite> attribute of the fiber view.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – Fiber view object containing k-mer count data in the ‘kmers’ element of the <cite>obsm</cite> attribute.</p></li>
<li><p><strong>metric</strong> (<em>str</em><em>, </em><em>optional</em>) – Distance metric to use for calculating pairwise distances. The default is ‘cityblock’.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.count_kmers">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">count_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.count_kmers" title="Link to this definition"></a></dt>
<dd><p>Count k-mers in each fiber in a fiber view.</p>
<p>This function counts the occurrences of k-mers in each fiber in a fiber view, and stores the resulting k-mer counts in the ‘kmers’ element of the <cite>obsm</cite> attribute of the fiber view. The length of the k-mers (<cite>k</cite>) and the mapping from k-mer strings to column indices in the k-mer count matrix are stored in the ‘kmer_len’ and ‘kmer_idx’ elements of the <cite>uns</cite> attribute, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – Fiber view object containing DNA sequence data in the ‘seq’ element of the <cite>layers</cite> attribute.</p></li>
<li><p><strong>k</strong> (<em>int</em>) – Length of the k-mers to count.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function updates the fiber view object in place, adding a new observation matrix ‘kmers’ containing
the counts of each k-mer for each fiber, and adds two new entries to the ‘uns’ dictionary: ‘kmer_len’ and
‘kmer_idx’. ‘kmer_len’ is the length of the k-mers that were counted, and ‘kmer_idx’ is a list of the
k-mers that were counted, with each k-mer represented as a bytes object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.filter_regions">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">filter_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nuc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_base_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length_limits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-inf,</span> <span class="pre">inf)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_limits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-inf,</span> <span class="pre">inf)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.filter_regions" title="Link to this definition"></a></dt>
<dd><p>Filter base modifications in a fiber view by length and score limits.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – The fiber view object containing the base modification data.</p></li>
<li><p><strong>base_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the type of regions to bin. This should be one of ‘nuc’ (nucleosomes), or ‘msp’ (methylation sensitive patches).
The default value is ‘nuc’.</p></li>
<li><p><strong>new_base_name</strong> (<em>str</em><em>, </em><em>optional</em>) – If not <cite>None</cite>, the new region base name to save the filtered regions to (region information at base_name will not be modified).
If new_base_name is None, the filtered regions will be saved to base_name.</p></li>
<li><p><strong>length_limits</strong> (<em>tuple</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – The lower and upper limits for the length of the base modifications. Modifications with lengths outside of these
limits will be filtered out. The default value is (-inf, inf), which includes all modifications.</p></li>
<li><p><strong>score_limits</strong> (<em>tuple</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – The lower and upper limits for the score of the base modifications. Modifications with scores outside of these
limits will be filtered out. The default value is (-inf, inf), which includes all modifications.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the function will filter the base modifications in place and return None. If False (default), the
function will return a new fiber view.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function updates the fiber view object in place or returns a new fiber
view with the selected region type filtered.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.get_seq_records">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">get_seq_records</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'read_name'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.get_seq_records" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.get_sequences">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">get_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.get_sequences" title="Link to this definition"></a></dt>
<dd><p>Returns a list of strings where each string is the sequence of one row of the fview object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>fview</strong> (<em>AnnData object</em>) – The fiber view object containing the sequence data.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>sequences</strong> – A list of strings where each string is the sequence of one row of the fview object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.make_dense_regions">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">make_dense_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nuc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ones'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.make_dense_regions" title="Link to this definition"></a></dt>
<dd><p>Create a dense matrix containing a representation of region infromation in a fiber view.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – The fiber view object containing the base modification data.</p></li>
<li><p><strong>base_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the type of regions to bin. This should be one of ‘nuc’ (nucleosomes), or ‘msp’ (methylation sensitive patches).
The default value is ‘nuc’.</p></li>
<li><p><strong>report</strong> (<em>str</em><em>, </em><em>optional</em>) – The data to include in the dense matrix. This should be one of ‘ones’, ‘score’ or ‘length’. The default value is ‘score’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dense matrix of size (number of fibers, number of bases) containing the specified region data.
Each position in the matrix where a region is not present is set to 0, positions where ar region is present
may be set to either the length or score value of the region occupying that position.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.make_region_df">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">make_region_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nuc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'left'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.make_region_df" title="Link to this definition"></a></dt>
<dd><p>Create a dataframe containing the positions and lengths of regions in a fiber view.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – The fiber view object containing the base modification data.</p></li>
<li><p><strong>base_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the type of regions to bin. This should be one of ‘nuc’ (nucleosomes), or ‘msp’ (methylation sensitive patches).
The default value is ‘nuc’.</p></li>
<li><p><strong>zero_pos</strong> (<em>str</em><em>, </em><em>optional</em>) – The position to use as the zero point for the start positions of the base modifications. This should be one of
‘left’, ‘center’, or ‘right’. The default value is ‘left’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dataframe with columns ‘row’ (the fiber index), ‘start’ (the start position of the base modification),
‘length’ (the length of the base modification), and ‘score’ (the score of the base modification).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.mark_cpg_sites">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">mark_cpg_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.mark_cpg_sites" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.plot_methylation">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">plot_methylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_bases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.plot_methylation" title="Link to this definition"></a></dt>
<dd><p>Plot a heatmap of DNA methylation levels in a fiber view.</p>
<p>This function generates a heatmap of DNA methylation levels in a fiber view, with different colors representing different methylation states (unmethylated, m6A, and 5mC). The DNA sequence for each fiber can optionally be labeled on the plot by setting the <cite>label_bases</cite> parameter to <cite>True</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – Fiber view object containing DNA methylation data in the ‘m6a’ and ‘cpg’ elements of the <cite>layers</cite> attribute, and DNA sequence data in the ‘seq’ element of the <cite>layers</cite> attribute.</p></li>
<li><p><strong>label_bases</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to label each base in the DNA sequence on the plot. The default is <cite>False</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ax</strong> – Axes object for the plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.axes._subplots.AxesSubplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.plot_summary">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">plot_summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sdata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.plot_summary" title="Link to this definition"></a></dt>
<dd><p>Plot a summary of DNA methylation levels and frequencies in a fiber view.</p>
<p>This function generates a line plot of DNA methylation levels and frequencies in a fiber view, with different lines representing different methylation states (m6A and 5mC). The data is summarized by grouping it into bins of a specified width, which can be set using the <cite>bin_width</cite> parameter (default is 10). The plot shows the total number of m6A and 5mC sites, as well as the frequencies of these modifications within each bin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sdata</strong> (<em>anndata.AnnData</em>) – Fiber view object containing DNA methylation data in the ‘m6a’ and ‘cpg’ elements of the <cite>layers</cite> attribute, and DNA sequence data in the ‘seq’ element of the <cite>layers</cite> attribute.</p></li>
<li><p><strong>bin_width</strong> (<em>int</em><em>, </em><em>optional</em>) – Width of the bins used to summarize the data. The default is 10.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ax</strong> – Axes object for the plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.axes._subplots.AxesSubplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.simple_region_plot">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">simple_region_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'m6a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.simple_region_plot" title="Link to this definition"></a></dt>
<dd><p>Plot a heatmap of DNA methylation levels and DNA sequence features in a fiber view.</p>
<p>This function generates a heatmap of DNA methylation levels and DNA sequence features in a fiber view. The data is displayed using a color map that indicates the presence or absence of different features (unmodified, nucleosome, m6A, m6A+nucleosome, msp, msp+m6A). The methylation state to be plotted can be specified using the <cite>mod</cite> parameter, which should be set to ‘m6a’ (default) or ‘cpg’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fview</strong> (<em>anndata.AnnData</em>) – Fiber view object containing DNA methylation data in the ‘m6a’ and ‘cpg’ elements of the <cite>layers</cite> attribute, and DNA sequence data in the ‘seq’ element of the <cite>layers</cite> attribute.</p></li>
<li><p><strong>mod</strong> (<em>str</em><em>, </em><em>optional</em>) – DNA methylation state to plot. Should be set to ‘m6a’ (default) or ‘cpg’.</p></li>
<li><p><strong>split_var</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of a field in fview.obs. if not None, a white line will be drawn between any rows where this value changes. Usefull for splitting clusters or groups of fibers up visually.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ax</strong> – Axes object for the plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.axes._subplots.AxesSubplot</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.tools.split_fire">
<span class="sig-prename descclassname"><span class="pre">fiber_views.tools.</span></span><span class="sig-name descname"><span class="pre">split_fire</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fview</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'msp'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['lnk',</span> <span class="pre">'fire']</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.tools.split_fire" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="fiber-views-utils-module">
<h2>fiber_views.utils module<a class="headerlink" href="#fiber-views-utils-module" title="Link to this heading"></a></h2>
<p id="module-fiber_views.utils">Created on Tue Aug 30 16:05:34 2022</p>
<p>&#64;author: morgan</p>
<dl class="py class">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">ReadList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">normal_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>A simple list of pysam.libcalignedsegment.PileupRead objects, plus methods
useful for constructing anndata elements from the read objects. also tracks
strand info of the genomic query position.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>normal_list</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of pysam.libcalignedsegment.PileupRead objects. The default is [].</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – The strand of the genomic query position. The default is “+”.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.strand">
<span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#fiber_views.utils.ReadList.strand" title="Link to this definition"></a></dt>
<dd><p>The strand of the genomic query position.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.build_anno_df">
<span class="sig-name descname"><span class="pre">build_anno_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anno_series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['np',</span> <span class="pre">'ec',</span> <span class="pre">'rq']</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.build_anno_df" title="Link to this definition"></a></dt>
<dd><p>Create a data frame with annotation data for the reads in the <cite>ReadList</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>anno_series</strong> (<em>pandas Series</em>) – A pandas Series containing annotation data for the genomic query position.</p></li>
<li><p><strong>tags</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of tags to include in the data frame. The default is [‘np’, ‘ec’, ‘rq’].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – A data frame with annotation data for the reads in the <cite>ReadList</cite> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.build_mod_array">
<span class="sig-name descname"><span class="pre">build_mod_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[('A',</span> <span class="pre">0,</span> <span class="pre">'a'),</span> <span class="pre">('T',</span> <span class="pre">1,</span> <span class="pre">'a')]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.build_mod_array" title="Link to this definition"></a></dt>
<dd><p>Create a base modification matrix for the reads in the <cite>ReadList</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> (<em>tuple</em>) – A tuple of integers representing the window of +/- window_offset. The
tuple should be of the form (window_start, window_end).</p></li>
<li><p><strong>mod_type</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of tuples representing the base modification type to consider.
The default is M6A_MODS.</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – The strand of the genomic query position. If not provided, the strand
information of the <cite>ReadList</cite> object is used. The default is None.</p></li>
<li><p><strong>sparse</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the base odification matrix is returned in sparse format. If
False, the matrix is returned in dense format. The default is True.</p></li>
<li><p><strong>score_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum score required for a base modification to be considered. The
default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mod_mtx</strong> – A base modification matrix for the reads in the <cite>ReadList</cite> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy array or scipy sparse matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.build_mod_array_from_def">
<span class="sig-name descname"><span class="pre">build_mod_array_from_def</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_def</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.build_mod_array_from_def" title="Link to this definition"></a></dt>
<dd><p>Create a base modification matrix for the reads in the <cite>ReadList</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> (<em>tuple</em>) – A tuple of integers representing the window of +/- window_offset. The
tuple should be of the form (window_start, window_end).</p></li>
<li><p><strong>mod_type</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of tuples representing the base modification type to consider.
The default is M6A_MODS.</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – The strand of the genomic query position. If not provided, the strand
information of the <cite>ReadList</cite> object is used. The default is None.</p></li>
<li><p><strong>sparse</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the base odification matrix is returned in sparse format. If
False, the matrix is returned in dense format. The default is True.</p></li>
<li><p><strong>score_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum score required for a base modification to be considered. The
default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mod_mtx</strong> – A base modification matrix for the reads in the <cite>ReadList</cite> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy array or scipy sparse matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.build_seq_array">
<span class="sig-name descname"><span class="pre">build_seq_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.build_seq_array" title="Link to this definition"></a></dt>
<dd><p>Create a byte array of the sequences for the reads in the <cite>ReadList</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> (<em>tuple</em>) – A tuple of integers representing the window of +/- window_offset. The
tuple should be of the form (window_start, window_end).</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – The strand of the genomic query position. If not provided, the strand
information of the <cite>ReadList</cite> object is used. The default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>char_array</strong> – A byte array of the sequences for the reads in the <cite>ReadList</cite> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy array</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Make sure to filter the reads in the <cite>ReadList</cite> object before using this
method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.build_sparse_region_array">
<span class="sig-name descname"><span class="pre">build_sparse_region_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('ns',</span> <span class="pre">'nl')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.build_sparse_region_array" title="Link to this definition"></a></dt>
<dd><p>Create a sparse region matrix for the reads in the <cite>ReadList</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> (<em>tuple</em>) – A tuple of integers representing the window of +/- window_offset. The
tuple should be of the form (window_start, window_end).</p></li>
<li><p><strong>tags</strong> (<em>tuple</em><em>, </em><em>optional</em>) – A tuple of tags to consider. The default is (‘ns’, ‘nl’).</p></li>
<li><p><strong>interval</strong> (<em>int</em><em>, </em><em>optional</em>) – The interval at which to report region information. This determines
the minimum window size that can be subset to that still preserve
region info. The default is 30.</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – The strand of the genomic query position. If not provided, the strand
information of the <cite>ReadList</cite> object is used. The default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>region_mtx</strong> – A sparse region matrix for the reads in the <cite>ReadList</cite> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>scipy sparse matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.filter_by_end_meth">
<span class="sig-name descname"><span class="pre">filter_by_end_meth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.filter_by_end_meth" title="Link to this definition"></a></dt>
<dd><p>Remove reads if they have fewer than [cutoff] m6A mods within [dist] base
pairs of the read start and end.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dist</strong> (<em>int</em><em>, </em><em>optional</em>) – The distance in base pairs from the read start and end to consider.
The default is 3000.</p></li>
<li><p><strong>cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum number of m6A mods within [dist] base pairs of the read
start and end required to keep the read. The default is 2.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the <cite>ReadList</cite> object is modified in place. If False, a new
<cite>ReadList</cite> object is returned. The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>inplace</cite> is True, returns None. If <cite>inplace</cite> is False, returns
a new <cite>ReadList</cite> object with the filtered reads and strand information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or <a class="reference internal" href="#fiber_views.utils.ReadList" title="fiber_views.utils.ReadList">ReadList</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.filter_by_window">
<span class="sig-name descname"><span class="pre">filter_by_window</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.filter_by_window" title="Link to this definition"></a></dt>
<dd><p>Remove reads that do not fully span a given window of +/- window_offset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> (<em>tuple</em>) – A tuple of integers representing the window of +/- window_offset. The
tuple should be of the form (window_start, window_end).</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the <cite>ReadList</cite> object is modified in place. If False, a new
<cite>ReadList</cite> object is returned. The default is False.</p></li>
<li><p><strong>strand</strong> (<em>str</em><em>, </em><em>optional</em>) – The strand of the genomic query position. If not provided, the strand
information of the <cite>ReadList</cite> object is used. The default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>inplace</cite> is True, returns None. If <cite>inplace</cite> is False, returns
a new <cite>ReadList</cite> object with the filtered reads and strand information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or <a class="reference internal" href="#fiber_views.utils.ReadList" title="fiber_views.utils.ReadList">ReadList</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.get_reads">
<span class="sig-name descname"><span class="pre">get_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_reads</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.get_reads" title="Link to this definition"></a></dt>
<dd><p>Retrieve reads from a BAM file for a given reference position.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment_file</strong> (<em>pysam.AlignmentFile</em>) – A BAM file opened with pysam.</p></li>
<li><p><strong>ref_pos</strong> (<em>tuple</em>) – A tuple containing the reference name, position, and strand of the
genomic query position. The tuple should be of the form
(reference_name, position, strand).</p></li>
<li><p><strong>max_reads</strong> (<em>int</em>) – the max number of reads to load from the bam file, usefull to speed
up processing when coverage is deep.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>self</strong> – The <cite>ReadList</cite> object with the reads and strand information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#fiber_views.utils.ReadList" title="fiber_views.utils.ReadList">ReadList</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>reads = ReadList().get_reads(bamfile, (‘chr3’, 200000, ‘+’))</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fiber_views.utils.ReadList.print_aligned_centers">
<span class="sig-name descname"><span class="pre">print_aligned_centers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.ReadList.print_aligned_centers" title="Link to this definition"></a></dt>
<dd><p>Print the center positions of the reads in the <cite>ReadList</cite> object with a specified number of bases on either side.
This is a test function to make sure reads are aligning correctly</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of bases on either side of the center position to include in the output. The default is 5.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.bed_to_anno_df">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">bed_to_anno_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry_name_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_id'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.bed_to_anno_df" title="Link to this definition"></a></dt>
<dd><p>Convert a data frame in BED format to another data frame with a different layout.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bed_df</strong> (<em>pandas.DataFrame</em>) – Data frame in BED format, with columns ‘chrom’, ‘start’, ‘end’, ‘strand’, ‘name’, and ‘score’.</p></li>
<li><p><strong>entry_name_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name for the unique identifier for each feature. The default is “gene_id”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>anno_df</strong> – Data frame with columns ‘seqid’, ‘pos’, ‘strand’, ‘entry_name_type’, and ‘score’.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.get_mod_pos_from_rec">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">get_mod_pos_from_rec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mods</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[('A',</span> <span class="pre">0,</span> <span class="pre">'a'),</span> <span class="pre">('T',</span> <span class="pre">1,</span> <span class="pre">'a')]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.get_mod_pos_from_rec" title="Link to this definition"></a></dt>
<dd><p>Retrieve positions of modified bases in a record.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rec</strong> (<em>pysam.libcalignedsegment.AlignedSegment</em>) – A record containing modified bases.</p></li>
<li><p><strong>mods</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of modified bases to consider, in the form (base, index, code).
The default is M6A_MODS.</p></li>
<li><p><strong>score_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum score required for a modified base to be included.
The default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mod_positions</strong> – An array of positions of modified bases.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>mod_positions = get_mod_pos_from_rec(read.alignment)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.get_strand_correct_mods">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">get_strand_correct_mods</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[('A',</span> <span class="pre">0,</span> <span class="pre">'a'),</span> <span class="pre">('T',</span> <span class="pre">1,</span> <span class="pre">'a')]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.get_strand_correct_mods" title="Link to this definition"></a></dt>
<dd><p>Retrieve modified bases in a read and correct their positions to match the forward genomic strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read</strong> (<em>pysam.libcalignedsegment.AlignedSegment</em>) – A read containing modified bases.</p></li>
<li><p><strong>mod_type</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of modified bases to consider, in the form (base, index, code).
The default is M6A_MODS.</p></li>
<li><p><strong>centered</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to center the positions around the query position of the read.
The default is False.</p></li>
<li><p><strong>score_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum score required for a modified base to be included.
The default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mods</strong> – An array of positions of modified bases, corrected for strand.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>mods = get_strand_correct_mods(read)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.get_strand_correct_mods_from_def">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">get_strand_correct_mods_from_def</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_def</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.get_strand_correct_mods_from_def" title="Link to this definition"></a></dt>
<dd><p>Retrieve modified bases in a read and correct their positions to match the forward genomic strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read</strong> (<em>pysam.libcalignedsegment.AlignedSegment</em>) – A read containing modified bases.</p></li>
<li><p><strong>mod_type</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of modified bases to consider, in the form (base, index, code).
The default is M6A_MODS.</p></li>
<li><p><strong>centered</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to center the positions around the query position of the read.
The default is False.</p></li>
<li><p><strong>score_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum score required for a modified base to be included.
The default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mods</strong> – An array of positions of modified bases, corrected for strand.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>mods = get_strand_correct_mods(read)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.get_strand_correct_regions">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">get_strand_correct_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('ns',</span> <span class="pre">'nl')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.get_strand_correct_regions" title="Link to this definition"></a></dt>
<dd><p>Retrieve start positions, lengths, and scores of regions in a read and correct them to match the forward genomic strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read</strong> (<em>pysam.libcalignedsegment.AlignedSegment</em>) – A read containing regions.</p></li>
<li><p><strong>tags</strong> (<em>tuple</em><em>, </em><em>optional</em>) – A tuple of tags containing the start positions, lengths, and scores of the
regions. The default is (‘ns’, ‘nl’).</p></li>
<li><p><strong>centered</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to center the positions around the query position of the read.
The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>starts</strong> (<em>numpy.ndarray</em>) – An array of start positions of regions, corrected for strand.</p></li>
<li><p><strong>lengths</strong> (<em>numpy.ndarray</em>) – An array of lengths of regions.</p></li>
<li><p><strong>scores</strong> (<em>numpy.ndarray</em>) – An array of scores of regions.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>starts, lengths, scores = get_strand_correct_regions(read)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.make_sparse_regions">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">make_sparse_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.make_sparse_regions" title="Link to this definition"></a></dt>
<dd><p>Make a sparse matrix representing genomic regions.</p>
<p>This function takes a DataFrame containing region information, as well as the
shape of the resulting matrix and other parameters, and returns three sparse
matrices representing the positions, lengths, and scores of the regions within
the matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region_df</strong> (<em>pandas.DataFrame</em>) – A DataFrame containing the region information. The DataFrame should have
columns for <cite>row</cite>, <cite>start</cite>, <cite>length</cite>, and <cite>score</cite>, representing the row
index of the matrix, the starting position of the region (0-based), the
length of the region, and the score associated with the region, respectively.</p></li>
<li><p><strong>shape</strong> (<em>tuple</em>) – The shape of the resulting matrix before binning. The first element should be the number
of rows in the matrix, and the second element should be the number of
columns.</p></li>
<li><p><strong>bin_width</strong> (<em>int</em><em>, </em><em>optional</em>) – The width of each bin in the resulting matrix, in base pairs. The default
is 1.</p></li>
<li><p><strong>interval</strong> (<em>int</em><em>, </em><em>optional</em>) – The interval at which to report region information. This determines the
minimum window size that can be subset to that still preserve region info.
The default is 30. interval is in number of bins, not bp</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing three sparse matrices representing the positions,
lengths, and scores of the regions within the matrix. The matrices are in
the form of COO sparse matrices. position values are still in base pairs
after binning. and may be negative for the first reported pos of a region</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of scipy.sparse.coo_matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.print_mod_contexts">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">print_mod_contexts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_positions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.print_mod_contexts" title="Link to this definition"></a></dt>
<dd><p>Print the contexts surrounding modified bases in a read.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>read</strong> (<em>pysam.libcalignedsegment.AlignedSegment</em>) – A read containing modified bases.</p></li>
<li><p><strong>mod_positions</strong> (<em>numpy.ndarray</em>) – An array of positions of modified bases in the read.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of bases on either side of the modified base to include in
the context. The default is 5.</p></li>
<li><p><strong>use_strand</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use the strand information in the read to determine the
context. If True, the context will be reversed if the read is on the
negative strand. The default is True.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<p>print_mod_contexts(read, mod_positions)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fiber_views.utils.read_bed">
<span class="sig-prename descclassname"><span class="pre">fiber_views.utils.</span></span><span class="sig-name descname"><span class="pre">read_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bed_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fiber_views.utils.read_bed" title="Link to this definition"></a></dt>
<dd><p>Read a BED file and return a pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>bed_file</strong> (<em>str</em>) – The file path of the BED file to be read.
The bed file should follow bed standard and not include column names.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the data from the BED file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-contents">
<h2>Module contents<a class="headerlink" href="#module-contents" title="Link to this heading"></a></h2>
<p id="module-fiber_views">Top-level package for fiber_views.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="examples.html" class="btn btn-neutral float-left" title="Examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Morgan Hamm.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>